web/src/LogStore.ts

```diff
@@ -0,0 +1,191 @@
+// Client-side log store, which helps client side rendering and filtering of logs.
+//
+// Looslely adapted from the data structures in
+// pkg/model/logstore/logstore.go
+// but with better support for incremental updates and rendering.
+
+type LogSpan = {
+  manifestName: string
+  lastSegmentIndex: number
+  firstSegmentIndex: number
+}
+
+class LogSegment {
+  spanId: string
+  time: string
+  text: string
+  continuesLine: boolean
+
+  constructor(seg: Proto.webviewLogSegment) {
+    this.spanId = seg.spanId || ""
```

Comment:
We can use `??` here now
```suggestion
    this.spanId = seg.spanId ?? ""
```

ditto on the ones below

Reply:
done